rocky/reports/report_types/tls_report/report.py

```diff
@@ -17,33 +19,58 @@ class TLSReport(Report):
     template_path = "tls_report/report.html"
     label_style = "3-light"
 
     def generate_data(self, input_ooi: str, valid_time: datetime) -> dict[str, Any]:
-        suites: dict = {}
-        findings: list[Finding] = []
-        suites_with_findings = []
-        ref = Reference.from_str(input_ooi)
-        tree = self.octopoes_api_connector.get_tree(
-            ref, valid_time=valid_time, depth=TREE_DEPTH, types={TLSCipher, Finding}
-        ).store
-        for pk, ooi in tree.items():
-            if ooi.ooi_type == "TLSCipher":
-                suites = ooi.suites
-            if ooi.ooi_type == "Finding" and ooi.finding_type.tokenized.id in CIPHER_FINDINGS:
-                findings.append(ooi)
-
-        for protocol, cipher_suites in suites.items():
-            for suite in cipher_suites:
-                for finding in findings:
-                    if finding.description and suite["cipher_suite_name"] in finding.description:
-                        suites_with_findings.append(suite["cipher_suite_name"])
-
-        return {
-            "input_ooi": input_ooi,
-            "suites": suites,
-            "findings": findings,
-            "suites_with_findings": suites_with_findings,
-        }
+        results = {}
+        class_type = Reference.from_str(input_ooi).class_type
+
+        if class_type == Hostname:
+            services = self.octopoes_api_connector.query(
+                "Hostname.<hostname[is ResolvedHostname].address.<address[is IPPort].<ip_port [is IPService]",
+                valid_time,
+                input_ooi,
+            )
+            oois = [ooi.primary_key for ooi in services if ooi.ooi_type == "IPService"]
+        elif class_type == IPAddressV4:
+            services = self.octopoes_api_connector.query(
+                "IPAddressV4.<address[is IPPort].<ip_port [is IPService]", valid_time, input_ooi
+            )
+            oois = [ooi.primary_key for ooi in services if ooi.ooi_type == "IPService"]
+        elif class_type == IPAddressV6:
+            services = self.octopoes_api_connector.query(
+                "IPAddressV6.<address[is IPPort].<ip_port [is IPService]", valid_time, input_ooi
+            )
+            oois = [ooi.primary_key for ooi in services if ooi.ooi_type == "IPService"]
+        else:
+            oois = [input_ooi]
+
+        for service in oois:
+            suites: dict = {}
+            findings: list[Finding] = []
+            suites_with_findings = []
+            ref = Reference.from_str(service)
+            tree = self.octopoes_api_connector.get_tree(
+                ref, valid_time=valid_time, depth=TREE_DEPTH, types={TLSCipher, Finding}
+            ).store
+            for pk, ooi in tree.items():
+                if ooi.ooi_type == "TLSCipher":
+                    suites = ooi.suites
+                if ooi.ooi_type == "Finding" and ooi.finding_type.tokenized.id in CIPHER_FINDINGS:
+                    findings.append(ooi)
```

Comment:
Could we do an actual query instead of a tree call here? It seems wasteful to fetch all these trees and filter them in python time.

Reply:
Yes, you are right.